Add Custom gateway tutorial

TucksonDev opened the pull request:
This tutorial creates a custom gateway on L1 and L2, as well as custom tokens for L1 and L2.
It demonstrates how to use the custom-custom way for bridging tokens.

gzeoneth commented on packages/custom-gateway-bridging/contracts/L1CustomGateway.sol:
```diff
@@ -0,0 +1,222 @@
+    address public l2CustomToken;
+    address public l2Gateway;
+    address public router;
+    bool private tokenBridgeInformationSet = false;
+
+    // Custom functionality
+    bool public allowsDeposits = false;
```
while this is safe bcz false is the default value, setting value when defining storage var can be risky when you use a proxy as it won't be set

TucksonDev replied:
This version is not prepared for a proxy, but in order to minimize the changes to be made to use a proxy, I've moved the default value to the constructor.

gzeoneth commented on packages/custom-gateway-bridging/contracts/L1CustomGateway.sol:
```diff
@@ -0,0 +1,222 @@
+        uint256 maxGas,
+        uint256 gasPriceBid,
+        bytes calldata data
+    ) public payable override returns (bytes memory res) {
```
you have a named return but ended with `return abi.encode(seqNum);`

TucksonDev commented on packages/custom-gateway-bridging/contracts/CrosschainMessenger.sol:
```diff
@@ -0,0 +1,161 @@
+/**
+ * @title Minimum expected implementation of a crosschain messenger contract to be deployed on L1
+ */
+abstract contract L1CrosschainMessenger {
```
token-bridge-contracts is currently in v0.6, and depend on openzeppelin/contracts v3.4.2, making it difficult to integrate with the current implementations.
As discussed, I'll continue with this version and update the tutorial when a new token-bridge-contracts is released.

TucksonDev commented:
Thanks @gvladika  and @gzeoneth !
I've updated the code including all your feedback (with the exception of importing the interfaces+abstract contracts from token-bridge-contracts, more info in that comment).
I've also added a test for withdrawals, and another one for the custom functionality (enable/disable deposits)

gvladika commented on packages/custom-gateway-bridging/contracts/L2CustomGateway.sol:
```diff
@@ -0,0 +1,210 @@
+
+        // Current exit number for this operation, after
+        // incrementing the exit number for future withdrawals
+        uint256 currExitNum = ++exitNum;
```
should be post-increment

gvladika reviewed (approved):
lgtm
